src/lib/operators/abstract_operator.hpp

```diff
@@ -216,20 +229,24 @@ class AbstractOperator : public std::enable_shared_from_this<AbstractOperator>,
   std::shared_ptr<const AbstractOperator> _left_input;
   std::shared_ptr<const AbstractOperator> _right_input;
 
-  bool _executed = false;
-  std::atomic<bool> _execution_started = false;
-
   // Is nullptr until the operator is executed
   std::shared_ptr<const Table> _output;
 
   // Weak pointer breaks cyclical dependency between operators and context
   std::optional<std::weak_ptr<TransactionContext>> _transaction_context;
 
+ private:
   // We track the number of consuming operators to automate the clearing of operator results.
-  std::atomic<int> _consumer_count = 0;
-
+  std::atomic_int32_t _consumer_count = 0;
   // Determines whether operator results can be cleared via clear_output().
   bool _never_clear_output = false;
+
+  enum class OperatorState : uint8_t { Created = 0, Running = 1, Executed = 2, Cleared = 3 };
+  std::atomic<OperatorState> _state{OperatorState::Created};
+  void _transition_to(OperatorState new_state);
+
+  std::weak_ptr<OperatorTask> _operator_task;
```

Comment:
Doc (1) what this is good for, (2) why this is must be a weak_ptr, (3) why we need a mutex, and (4) that this should be replaced with `atomic_weak_ptr` once we have it.

Reply:
Changed to

```cpp
  /**
   * OperatorTasks wrap operators for scheduling. Since operator results are shared between uncorrelated subqueries
   * and their outer queries, OperatorTasks should be shared, too, to reduce scheduling overhead and to prevent
   * additional logic for result sharing.
   * To allow OperatorTask::make_tasks_from_operator to reuse an existing OperatorTask, operators create
   * OperatorTasks from themselves and store weak pointers. The pointers must be weak because OperatorTasks also
   * point to operators, which would otherwise create cyclic dependencies.
   */
  std::weak_ptr<OperatorTask> _operator_task;
  // To prevent race conditions in get_or_create_operator_task.
  std::mutex _operator_task_mutex;
```

> (4) that this should be replaced with atomic_weak_ptr once we have it.

My assumption is, that we cannot abandon the mutex with an `atomic_weak_ptr`. Probably, we would end up with race conditions since we do more stuff after having created the OperatorTask:

```cpp
  auto operator_task = std::make_shared<OperatorTask>(shared_from_this());
  _operator_task = std::weak_ptr<OperatorTask>(operator_task);
  if (executed()) {
    // Skip task to reduce scheduling overhead.
    operator_task->skip_operator_task();
    DebugAssert(operator_task->is_done(), "Expected OperatorTask to be marked as done.");
  }

  return operator_task;
```